pymtl3/dsl/Component.py

```diff
@@ -55,10 +61,543 @@ def _construct( s ):
+
+  """ Convenience/utility APIs """
+  def apply( s, *args ):
+
+    if isinstance(args[0], list):
+      assert len(args) == 1
+      for step in args[0]:
+        step( s )
+
+    elif len(args) == 1:
+      assert callable( args[0] )
+      args[0]( s )
+
+  # Simulation related APIs
   def sim_reset( s ):
-    assert s._dsl.elaborate_top is s # assert sim_reset is top
+    s._check_called_at_elaborate_top( "sim_reset" )
 
     s.reset = Bits1( 1 )
     s.tick() # This tick propagates the reset signal
     s.reset = Bits1( 0 )
     s.tick()
+
+  def check( s ):
+    s._check_valid_dsl_code()
+
+  # TODO maybe we should implement these two lock/unlock APIs as passes?
+  # They expose kernel implementation details though ...
+  def lock_in_simulation( s ):
+    s._check_called_at_elaborate_top( "lock_in_simulation" )
+
+    s._dsl.swapped_signals = defaultdict(list)
+    s._dsl.swapped_values  = defaultdict(list)
+
+    def cleanup_connectables( current_obj, host_component ):
+
+      # Deduplicate code. Choose operation based on type of current_obj
+      if isinstance( current_obj, list ):
+        iterable = enumerate( current_obj )
+        is_list = True
+      elif isinstance( current_obj, NamedObject ):
+        iterable = current_obj.__dict__.iteritems()
+        is_list = False
+      else:
+        return
+
+      for i, obj in iterable:
+        if not is_list and i.startswith("_"): # impossible to have tuple
+          continue
+
+        if   isinstance( obj, Component ):
+          cleanup_connectables( obj, obj )
+
+        elif isinstance( obj, (Interface, list) ):
+          cleanup_connectables( obj, host_component )
+
+        elif isinstance( obj, Signal ):
+          try:
+            if is_list: current_obj[i] = obj.default_value()
+            else:       setattr( current_obj, i, obj.default_value() )
+          except Exception as err:
+            err.message = repr(obj) + " -- " + err.message
+            err.args = (err.message,)
+            raise err
+          s._dsl.swapped_signals[ host_component ].append( (current_obj, i, obj, is_list) )
+
+    cleanup_connectables( s, s )
+    s._dsl.locked_simulation = True
+
+  def unlock_simulation( s ):
+    s._check_called_at_elaborate_top( "unlock_simulation" )
+    try:
+      assert s._dsl.locked_simulation
+    except:
+      raise AttributeError("Cannot unlock an unlocked/never locked model.")
+
+    for component, records in s._dsl.swapped_signals.iteritems():
+      for current_obj, i, obj, is_list in records:
+        if is_list:
+          s._dsl.swapped_values[ component ] = ( current_obj, i, current_obj[i], is_list )
+          current_obj[i] = obj
+        else:
+          s._dsl.swapped_values[ component ] = ( current_obj, i, getattr(current_obj, i), is_list )
+          setattr( current_obj, i, obj )
+
+    s._dsl.locked_simulation = False
+
+  """ APIs that provide local metadata of a component """
```

Comment:
Same here. I feel our `#-----` style separator works well in these kinds of situations. What do you think?